lib/facade_converter.ts

```diff
@@ -0,0 +1,150 @@
+/// <reference path='../node_modules/typescript/bin/typescript.d.ts' />
+import base = require('./base');
+import ts = require('typescript');
+import ts2dart = require('./main');
+
+type FacadeHandler = (c: ts.CallExpression, context: ts.Expression) => void;
+
+export class FacadeConverter extends base.TranspilerStep {
+  private tc: ts.TypeChecker;
+  private forbiddenNames: {[fileName: string]: boolean};
+
+  constructor(transpiler: ts2dart.Transpiler) {
+    super(transpiler);
+    this.forbiddenNames = {};
+    for (var fileName in this.subs) {
+      Object.keys(this.subs[fileName])
+          .map((fnName) => fnName.substring(fnName.lastIndexOf('.') + 1))
+          .forEach((fnName) => this.forbiddenNames[fnName] = true);
+    }
+  }
+
+  setTypeChecker(tc: ts.TypeChecker) { this.tc = tc; }
+
+  handleCall(c: ts.CallExpression): boolean {
+    if (!this.tc) return false;
+    var symbol: ts.Symbol;
+    var context: ts.Expression;
+    if (c.expression.kind === ts.SyntaxKind.Identifier) {
+      // Function call.
+      symbol = this.tc.getSymbolAtLocation(c.expression);
+      context = null;
+    } else if (c.expression.kind === ts.SyntaxKind.PropertyAccessExpression) {
+      // Method call.
+      var pa = <ts.PropertyAccessExpression>c.expression;
+      symbol = this.tc.getSymbolAtLocation(pa.name);
+      context = pa.expression;
+    } else {
+      return false;
+    }
+
+    if (!symbol) return false;
+    if (symbol.flags & ts.SymbolFlags.Alias) symbol = this.tc.getAliasedSymbol(symbol);
+    if (!symbol.valueDeclaration) return false;
+
+    var fileName = symbol.valueDeclaration.getSourceFile().fileName;
+    fileName = fileName.replace(/(\.d)?\.ts$/, '');
+    // console.log('fn:', fileName);
+    var fileSubs = this.subs[fileName];
+    if (!fileSubs) return false;
+
+    var qn = this.tc.getFullyQualifiedName(symbol);
+    // console.log('qn', qn);
+    var qnSub = fileSubs[qn];
+    if (!qnSub) return false;
+
+    qnSub(c, context);
+    return true;
+  }
+
+  private subs: ts.Map<ts.Map<FacadeHandler>> = {
+    'lib': {
+      'Array.push': (c: ts.CallExpression, context: ts.Expression) => {
```

Comment:
can type inference work here, so you could just `(c, context) => {}`
would make this structure more readable.